src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,375 @@
+        auto maybeFn = tryGetFn(lock, entrypointName, ctx, methodName);
+        KJ_IF_SOME(e, maybeFn.tryGet<kj::String>()) {
+          exceptionThrown = true;
+          reader.setV8Serialized(_::serializeV8(js, js.error(e)));
+          co_return exceptionThrown;
+        }
+
+        auto fn = KJ_REQUIRE_NONNULL(maybeFn.tryGet<v8::Local<v8::Function>>());
+        kj::Maybe<SerializedResult> eventualResult;
+
+        // We have a function, so let's call it, get the result, and serialize it for RPC.
+        // We account for both the JS method throwing and our serialization attempt failing.
+        if (fn->IsAsyncFunction()) {
+          eventualResult = co_await prepareAsyncResult(js, fn, serializedArgs);
+        } else {
+          eventualResult = prepareSyncResult(js, fn, serializedArgs);
         }
-    ));
+        auto result = kj::mv(KJ_REQUIRE_NONNULL(eventualResult));
+        exceptionThrown = result.caughtException;
+
+        // Note that returning an exception from JS will not set `exceptionThrown`.
+        reader.setV8Serialized(kj::mv(result.data));
+        co_return exceptionThrown;
+      });
+    } catch(kj::Exception e) {
+      caughtException = true;
+      if (auto desc = e.getDescription();
+          !jsg::isTunneledException(desc) && !jsg::isDoNotLogException(desc)) {
+        LOG_EXCEPTION("JsRpcTargetCall"_kj, e);
+      }
+    }
+    // Upon returning, we want to fulfill the callPromise so customEvent can continue executing.
+    // Recall that we inform the customEvent whether we completed cleanly, i.e. if we threw an
+    // exception.
+    KJ_DEFER(callFulfiller->fulfill(kj::mv(caughtException)););
+    co_return;
+  }
+
+  KJ_DISALLOW_COPY_AND_MOVE(JsRpcTargetImpl);
+
+private:
+  // The following names are registered by the Workers Runtime and cannot be called over RPC.
+  bool isRegisteredName(kj::StringPtr name) {
+    if (name == "constructor"_kjc ||
+        name == "fetch"_kjc ||
+        name == "alarm"_kjc ||
+        name == "webSocketMessage"_kjc ||
+        name == "webSocketClose"_kjc ||
+        name == "webSocketError"_kjc) {
+      return true;
+    }
+    return false;
+  }
+
+  // If the `methodName` is a known public method, we'll return it.
+  inline kj::OneOf<v8::Local<v8::Function>, kj::String> tryGetFn(
+      Worker::Lock& lock,
+      kj::Maybe<kj::StringPtr> entrypointName,
+      IoContext& ctx,
+      kj::StringPtr methodName) {
+
+    auto& handler = KJ_REQUIRE_NONNULL(lock.getExportedHandler(entrypointName, ctx.getActor()),
+        "Failed to get handler to worker.");
+
+    if (isRegisteredName(methodName)) {
+      return kj::str("'", methodName, "' is a registered method and cannot be called over RPC."_kj);
+    }
+    auto handle = handler.self.getHandle(lock);
+    auto methodStr = jsg::v8StrIntern(lock.getIsolate(), methodName);
+    auto fnHandle = jsg::check(handle->Get(lock.getContext(), methodStr));
+    if (!(fnHandle->IsFunction() && !fnHandle->IsPrivate())) {
```

Comment:
Does `Get()` actually return private properties in the first place? I'd kind of expect it wouldn't, making the `IsPrivate()` check redundant. Of course, this should be tested carefully.

Reply:
Ah, regarding private methods I added this comment to the named intercept PR a couple days back (cc @jasnell):

> Just noticed that if I define a private method on my DO and try to call it from the client side, I get `jsg.SyntaxError: Private field '#getPrivate' must be declared in an enclosing class` and `getNamed()` isn't invoked.
>
> Is there any way we can have private names also be intercepted? We want to reject calls to private methods over RPC, but I think it would be better if the script got a clear exception stating we don't allow calling private methods over RPC rather than the message above.

Reply:
Get does not handle private properties. You must use `GetPrivate()` (https://v8docs.nodesource.com/node-20.3/db/d85/classv8_1_1_object.html#afa05f42683ca4029b623721330b51113) for those

I do not believe it is possible to intercept private fields/methods using the interceptor mechanism currently as I do not believe v8 implements that. I'll investigate further tho

Reply:
https://github.com/cloudflare/workerd/compare/b249885946d71f10959e36108c3ee43d10858753..f12f95f6a9e2c4ce0fd51d7c21e54c46276ae140